libs/server/TLS/GarnetTlsOptions.cs

```diff
@@ -161,21 +161,69 @@ SslClientAuthenticationOptions GetSslClientAuthenticationOptions()
+            return (object _, X509Certificate certificate, X509Chain __, SslPolicyErrors sslPolicyErrors)
+                => (sslPolicyErrors == SslPolicyErrors.None) || (sslPolicyErrors == SslPolicyErrors.RemoteCertificateChainErrors
+                   && certificate is X509Certificate2 certificate2
+                   && ValidateCertificateName(certificate2, targetHostName)
+                   && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Validates certificate subject name by looking into DNS name property (preferred), if missing it falls back to
+        /// legacy SimpleName. The input certificate subject should match the expected host name provided in server config.
+        /// </summary>
+        /// <param name="certificate2"> remote certificate to validate.</param>
+        /// <param name="targetHostName"> target host name configured. </param>
+        private bool ValidateCertificateName(X509Certificate2 certificate2, string targetHostName)
+        {
+            var subjectName = certificate2.GetNameInfo(X509NameType.DnsName, false);
+            if (string.IsNullOrWhiteSpace(subjectName))
+            {
+                subjectName = certificate2.GetNameInfo(X509NameType.SimpleName, false);
+            }
+
+            return subjectName.Equals(targetHostName, StringComparison.InvariantCultureIgnoreCase);
+        }
+
         /// <summary>
         /// Callback to verify the TLS certificate
         /// </summary>
         /// <param name="issuerCertificatePath"></param>
         /// <returns></returns>
         /// <exception cref="ArgumentException"></exception>
         /// <exception cref="ArgumentNullException"></exception>
-        RemoteCertificateValidationCallback ValidateCertificateCallback(string issuerCertificatePath)
+        RemoteCertificateValidationCallback ValidateClientCertificateCallback(string issuerCertificatePath)
         {
             if (!ClientCertificateRequired)
             {
                 logger?.LogWarning("ClientCertificateRequired is false. Remote certificate validation will always succeed.");
                 return (object _, X509Certificate certificate, X509Chain __, SslPolicyErrors sslPolicyErrors)
                     => true;
             }
+            var issuer = GetCertIssuer(issuerCertificatePath);
+            return (object _, X509Certificate certificate, X509Chain __, SslPolicyErrors sslPolicyErrors)
+                => (sslPolicyErrors == SslPolicyErrors.None) || (sslPolicyErrors == SslPolicyErrors.RemoteCertificateChainErrors
+                    && certificate is X509Certificate2 certificate2
+                    && ValidateCertificateIssuer(certificate2, issuer));
+        }
+
+        /// <summary>
+        /// Returns issuer certificate in X509Certificate2 format.
+        /// </summary>
+        /// <param name="issuerCertificatePath">path to issuer cert.</param>
```

Comment:
NIT: I'd clean up the comment a bit.

```suggestion
        /// <summary>
        /// Loads an issuer X.509 certificate using its file name.
        /// </summary>
        /// <param name="issuerCertificatePath">The path of the certificate file.</param>
```